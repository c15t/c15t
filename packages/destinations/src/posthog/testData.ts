// cSpell:disable
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-nocheck
import type {
	ActionDefinition,
	DestinationDefinition,
	InputField,
} from '@segment/actions-core';

import Chance from 'chance';

type Choices = Pick<InputField, 'choices'>;

function setTestData(
	seedName: string,
	type: string,
	fieldName?: string,
	format?: string,
	choices?: Choices
) {
	const chance = new Chance(seedName);

	if (Array.isArray(choices)) {
		const choice = chance.pickone(choices);

		if (choice.value) {
			return choice.value;
		}
		return choice;
	}
	let val: any;
	switch (type) {
		case 'boolean':
			val = chance.bool();
			break;
		case 'datetime':
			val = '2021-02-01T00:00:00.000Z';
			break;
		case 'integer':
			val = chance.integer();
			break;
		case 'number':
			val = chance.floating({ fixed: 2 });
			break;
		case 'object':
			val = { testType: chance.string() };
			break;
		case 'text':
			val = chance.sentence();
			break;
		default:
			// covers string
			switch (format) {
				case 'date': {
					val = '2021-02-01T00:00:00.000Z';
					break;
				}
				case 'date-time':
					val = '2021-02-01T00:00:00.000Z';
					break;
				case 'email':
					val = chance.email();
					break;
				case 'hostname':
					val = chance.domain();
					break;
				case 'ipv4':
					val = chance.ip();
					break;
				case 'ipv6':
					val = chance.ipv6();
					break;
				case 'time': {
					const d = chance.date();
					val = [d.getHours(), d.getMinutes(), d.getSeconds()]
						.map((v) => String(v).padStart(2, '0'))
						.join(':');
					break;
				}
				case 'uri':
					val = chance.url();
					break;
				case 'uuid':
					val = chance.guid();
					break;
				default:
					val = chance.string();
					break;
			}
			break;
	}

	if (fieldName === 'email') {
		val = chance.email();
	}
	if (fieldName === 'currency') {
		val = chance.currency().code;
	}

	return val;
}

function setData(
	eventData: any,
	chanceName: string,
	fieldName: string,
	field: any,
	data?: any
) {
	const { format, multiple, type } = field;

	if (!data) {
		// biome-ignore lint/style/noParameterAssign: This is a generated file
		data = setTestData(chanceName, type, fieldName, format, field.choices);
	}

	eventData[fieldName] = multiple ? [data] : data;
	return eventData;
}

export function generateTestData(
	seedName: string,
	destination: DestinationDefinition<any>,
	action: ActionDefinition<any>,
	isRequiredOnly: boolean
) {
	let eventData: any = {};
	const settingsData: any = {};

	const authentication = destination.authentication;
	if (authentication) {
		for (const settingKey in authentication.fields) {
			const { format, type } = authentication.fields[settingKey];
			settingsData[settingKey] = setTestData(seedName, type, undefined, format);
		}
	}

	for (const [name, field] of Object.entries(action.fields)) {
		if (isRequiredOnly && !(field.required || name.includes('id'))) {
			continue;
		}

		const { properties } = field;

		if (properties) {
			let subData: any = {};
			let propertyFields = Object.keys(properties);

			if (isRequiredOnly) {
				propertyFields = propertyFields.filter(
					(name) => properties[name].required
				);
			}

			for (const propertyName of propertyFields) {
				const property = properties[propertyName];
				subData = setData(subData, seedName, propertyName, property);
			}

			eventData = setData(eventData, seedName, name, field, subData);
			continue;
		}

		eventData = setData(eventData, seedName, name, field);
	}

	return [eventData, settingsData];
}
