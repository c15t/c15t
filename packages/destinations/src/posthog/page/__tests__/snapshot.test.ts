import { createTestEvent, createTestIntegration } from '@segment/actions-core';
import nock from 'nock';
import { describe, expect, it } from 'vitest';

import destination from '../../index';
import { generateTestData } from '../../testData';

const testDestination = createTestIntegration(destination);
const actionSlug: keyof typeof destination.actions = 'page';
const destinationSlug = 'PostHog';
const seedName = `${destinationSlug}#${actionSlug}`;

describe(`Testing snapshot for ${destinationSlug}'s ${actionSlug} destination action:`, () => {
	it('required fields', async () => {
		const action = destination.actions[actionSlug];
		if (!action) {
			return null;
		}

		const [eventData, settingsData] = generateTestData(
			seedName,
			destination,
			action,
			true
		);

		nock(/.*/).persist().get(/.*/).reply(200);
		nock(/.*/).persist().post(/.*/).reply(200);
		nock(/.*/).persist().put(/.*/).reply(200);

		const event = createTestEvent({
			properties: eventData,
		});

		const responses = await testDestination.testAction(actionSlug, {
			auth: undefined,
			event: event,
			mapping: event.properties,
			settings: settingsData,
		});

		if (!responses[0]) {
			return null;
		}
		const request = responses[0].request;
		const rawBody = await request.text();

		const json = JSON.parse(rawBody);
		expect(json).toMatchSnapshot();
	});

	it('all fields', async () => {
		const action = destination.actions[actionSlug];
		if (!action) {
			return null;
		}

		const [eventData, settingsData] = generateTestData(
			seedName,
			destination,
			action,
			false
		);

		nock(/.*/).persist().get(/.*/).reply(200);
		nock(/.*/).persist().post(/.*/).reply(200);
		nock(/.*/).persist().put(/.*/).reply(200);

		const event = createTestEvent({
			properties: eventData,
		});

		const responses = await testDestination.testAction(actionSlug, {
			auth: undefined,
			event: event,
			mapping: event.properties,
			settings: settingsData,
		});

		if (!responses[0]) {
			return null;
		}

		const request = responses[0].request;
		const rawBody = await request.text();

		try {
			const json = JSON.parse(rawBody);
			expect(json).toMatchSnapshot();
			return;
		} catch {
			expect(rawBody).toMatchSnapshot();
		}
	});
});
