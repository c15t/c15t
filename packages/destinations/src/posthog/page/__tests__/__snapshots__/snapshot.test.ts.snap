// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing snapshot for PostHog's page destination action: > all fields 1`] = `
{
  "api_key": "5B(%pAE!!G",
  "distinct_id": "5B(%pAE!!G",
  "event": "$pageview",
  "headers": {},
  "properties": {
    "$browser": "5B(%pAE!!G",
    "$browser_language": "5B(%pAE!!G",
    "$browser_version": -31403816423260.16,
    "$current_url": "5B(%pAE!!G",
    "$device_type": "5B(%pAE!!G",
    "$host": "5B(%pAE!!G",
    "$insert_id": "5B(%pAE!!G",
    "$ip": "5B(%pAE!!G",
    "$lib": "5B(%pAE!!G",
    "$lib_version": "5B(%pAE!!G",
    "$os": "5B(%pAE!!G",
    "$os_version": "5B(%pAE!!G",
    "$pathname": "5B(%pAE!!G",
    "$raw_user_agent": "5B(%pAE!!G",
    "$referrer": "5B(%pAE!!G",
    "$referring_domain": "5B(%pAE!!G",
    "$screen_dpr": -31403816423260.16,
    "$screen_height": -31403816423260.16,
    "$screen_width": -31403816423260.16,
    "$session_id": "5B(%pAE!!G",
    "$time": -31403816423260.16,
    "$viewport_height": -31403816423260.16,
    "$viewport_width": -31403816423260.16,
    "$window_id": "5B(%pAE!!G",
    "title": "5B(%pAE!!G",
    "utm_campaign": "5B(%pAE!!G",
    "utm_content": "5B(%pAE!!G",
    "utm_medium": "5B(%pAE!!G",
    "utm_source": "5B(%pAE!!G",
    "utm_term": "5B(%pAE!!G",
  },
  "timestamp": "2021-02-01T00:00:00.000Z",
}
`;

exports[`Testing snapshot for PostHog's page destination action: > required fields 1`] = `
{
  "api_key": "5B(%pAE!!G",
  "distinct_id": "5B(%pAE!!G",
  "event": "$pageview",
  "headers": {},
  "properties": {
    "$insert_id": "5B(%pAE!!G",
    "$screen_width": -31403816423260.16,
    "$session_id": "5B(%pAE!!G",
    "$viewport_width": -31403816423260.16,
    "$window_id": "5B(%pAE!!G",
  },
  "timestamp": "2021-02-01T00:00:00.000Z",
}
`;
